### PR TITLE
Extract a ShardStateObserver from ShardStateAwareRemoteCollector

### PR DESCRIPTION
**Pull request by mfussenegger**
## Summary of the changes / Why this improves CrateDB

This extracts a `ShardStateObserver` from
`ShardStateAwareRemoteCollector`. This should separate the concerns of
the components better. `ShardStateObserver` is only responsible to wait
for a shard to become active again. Everything else moves into
  `RemoteCollectorFactory`.

## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by mfussenegger** [comment]
@kovrus maybe you can already have a look at this and let me know what you think of it. I'll add tests if you think this is an improvement.

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/ShardStateObserver.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.index.IndexNotFoundException;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.shard.ShardNotFoundException;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeoutException;
+
+public final class ShardStateObserver {
```
Comment (kovrus) [inline sql/src/main/java/io/crate/execution/engine/collect/collectors/ShardStateObserver.java]: more clear! nice separation of concerns!

File: sql/src/main/java/io/crate/execution/engine/collect/RemoteCollectorFactory.java [hunk 2]
```diff
@@ -98,73 +99,64 @@ public RemoteCollectorFactory(ClusterService clusterService,
...
             true
         );
     }
 
-    private Function<IndexShard, BatchIterator<Row>> getLocalCollectorProvider(ShardCollectorProviderFactory shardCollectorProviderFactory,
-                                                                               RoutedCollectPhase collectPhase,
-                                                                               CollectTask collectTask,
-                                                                               RowConsumer consumer) {
-        return indexShard -> {
+    private CompletableFuture<List<Row>> retrieveRows(ShardRouting activePrimaryRouting,
```
Comment (kovrus) [inline sql/src/main/java/io/crate/execution/engine/collect/RemoteCollectorFactory.java]: imho, it was a bit more readable and easier to grasp the logic immediately when we had `getLocalCollectorProvider` and `getRemoteCollectorProvider` methods

Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/collect/RemoteCollectorFactory.java]: Any suggestions how to restore that? The if/else blocks within `retrieveRows` are rather short, so I'm not sure if it makes sense to extract some methods to clarify that one case is remote and another local

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/ShardStateObserver.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+    private static final TimeValue MAX_WAIT_TIME_FOR_NEW_STATE = TimeValue.timeValueSeconds(30);
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/collect/collectors/ShardStateObserver.java]: Not sure. Request timeouts tend to be 30 seconds, so I changed it to be consistent with them.

File: sql/src/main/java/io/crate/execution/engine/collect/RemoteCollectorFactory.java (hunk 2 above)
Comment (kovrus) [reply sql/src/main/java/io/crate/execution/engine/collect/RemoteCollectorFactory.java]: thought of keeping those methods, but it doesn't really matter 👍

**Pull request merged**